viogpu/common/viogpu_idr.h

```diff
@@ -42,7 +43,16 @@ class VioGpuIdr
     VOID Lock(VOID);
     VOID Unlock(VOID);
 private:
-    ULONG m_uStartIndex;
-    RTL_BITMAP m_IdBitMap;
-    FAST_MUTEX m_IdBitMapMutex;
+    static const ULONG IdsBlockSize = 1000;
+
+    struct FreeIds {
+        LIST_ENTRY list_entry;
+        UINT Available;
+        ULONG FreeIDs[VioGpuIdr::IdsBlockSize];
+    };
+
+    ULONG m_nextId;
+    FAST_MUTEX m_mutex;
+    FreeIds *m_current;
+    LIST_ENTRY m_blocks;
```

Comment:
Checkout USBdk project for C++ wrapper for lists with locks and different pools allocations:
https://github.com/daynix/UsbDk/blob/af46b7ee3302413d2533c492fac2aa154cfd3d70/UsbDk/UsbDkUtil.h#L245

Reply:
Initially this complex queue was written because it was sitting in fairly hot place (submitting commands), now that i gave it a little bit more thought it can be replaced with just atomic counter ([like linux does](https://github.com/torvalds/linux/blob/d88520ad73b79e71e3ddf08de335b8520ae41c5c/drivers/gpu/drm/virtio/virtgpu_fence.c#L109)). This means queue will be used only in resource id allocation, which is much colder path and allows us to use simpler approach with interlocked lists. I will implement this soon.

Reply:
It is now fixed, command ID generation uses atomic counter and ID generation for resources and contexts uses `ExInterlocked*List` routines which simplified code a lot.